File: backend/application/question_quality.py
"""Framing quality gates shared by Ask / sequential start."""

from __future__ import annotations

import re
import unicodedata
from typing import Tuple

MIN_QUESTION_WORDS = 8
MIN_UNIQUE_WORDS = 5

_JUNK_ONLY = re.compile(
    r"^(test(ing)?|asdf|qwerty|lorem|ipsum|xxx|spam|foo|bar|baz|hello|hi|hey)"
    r"([,\s.!?-]+\1)*[.\s!?]*$",
    re.IGNORECASE,
)

_TOKEN = re.compile(r"\S+")


def _normalize_token(raw: str) -> str:
    text = unicodedata.normalize("NFKD", raw.lower())
    text = "".join(ch for ch in text if not unicodedata.combining(ch))
    return re.sub(r"^[^\w]+|[^\w]+$", "", text, flags=re.UNICODE)


def tokenize(text: str) -> list:
    return [t for t in (_normalize_token(m.group(0)) for m in _TOKEN.finditer(text or "")) if t]


def count_words(text: str) -> int:
    return len(tokenize(text))


def assess_question_quality(text: str) -> Tuple[bool, str]:
    cleaned = (text or "").strip()
    tokens = tokenize(cleaned)
    word_count = len(tokens)
    unique_count = len(set(tokens))

    if word_count < MIN_QUESTION_WORDS:
        return False, (
            f"Question needs at least {MIN_QUESTION_WORDS} words "
            f"(got {word_count})."
        )

    compact = re.sub(r"\s+", " ", cleaned)
    if _JUNK_ONLY.match(compact):
        return False, (
            "Framing looks like placeholder or test text — describe a real situation."
        )

    if unique_count < MIN_UNIQUE_WORDS:
        return False, (
            "Framing is too repetitive — use at least a few distinct words."
        )

    freq = {}
    for token in tokens:
        freq[token] = freq.get(token, 0) + 1
    top = max(freq.values()) if freq else 0
    if word_count and top / word_count > 0.4:
        return False, "Framing repeats the same words too often — add real context."

    return True, ""


def validate_question_framing(question: str) -> str:
    cleaned = (question or "").strip()
    ok, message = assess_question_quality(cleaned)
    if not ok:
        raise ValueError(message)
    return cleaned

File: frontend/src/utils/questionQuality.js
/** Framing quality gates — block empty, repeated junk, and non-questions from burning API calls. */

export const MIN_QUESTION_WORDS = 8;
export const MIN_UNIQUE_WORDS = 5;

const JUNK_ONLY =
  /^(test(ing)?|asdf|qwerty|lorem|ipsum|xxx|spam|foo|bar|baz|hello|hi|hey)([,\s.!?-]+\1)*[.\s!?]*$/i;

function normalizeToken(raw) {
  return raw
    .toLowerCase()
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/^[^\p{L}\p{N}]+|[^\p{L}\p{N}]+$/gu, "");
}

export function tokenize(text = "") {
  return (text.match(/\S+/g) || []).map(normalizeToken).filter(Boolean);
}

export function countWords(text = "") {
  return tokenize(text).length;
}

/**
 * @returns {{ ok: boolean, reason: string|null, wordCount: number, uniqueCount: number }}
 */
export function assessQuestionQuality(text = "") {
  const cleaned = text.trim();
  const tokens = tokenize(cleaned);
  const wordCount = tokens.length;
  const unique = new Set(tokens);
  const uniqueCount = unique.size;

  if (wordCount < MIN_QUESTION_WORDS) {
    return { ok: false, reason: "too_short", wordCount, uniqueCount };
  }

  if (JUNK_ONLY.test(cleaned.replace(/\s+/g, " "))) {
    return { ok: false, reason: "junk", wordCount, uniqueCount };
  }

  if (uniqueCount < MIN_UNIQUE_WORDS) {
    return { ok: false, reason: "repetitive", wordCount, uniqueCount };
  }

  // Dominant token > 40% of words → spam / filler
  const freq = {};
  for (const t of tokens) freq[t] = (freq[t] || 0) + 1;
  const top = Math.max(...Object.values(freq));
  if (top / wordCount > 0.4) {
    return { ok: false, reason: "repetitive", wordCount, uniqueCount };
  }

  return { ok: true, reason: null, wordCount, uniqueCount };
}

export function questionMeetsQuality(text = "") {
  return assessQuestionQuality(text).ok;
}
